File: iBurn/src/main/java/com/gaiagps/iburn/js/JSEvaluator.java
package com.gaiagps.iburn.js;

import android.content.Context;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import java.util.concurrent.atomic.AtomicBoolean;

import io.reactivex.android.schedulers.AndroidSchedulers;
import rx.Observable;
import rx.subjects.PublishSubject;
import timber.log.Timber;

/**
 * Created by dbro on 8/6/15.
 */
public class JSEvaluator {

    private static final PublishSubject jsEvaluatorSubject = PublishSubject.create();
    private static final AtomicBoolean initializedJsEvaluator = new AtomicBoolean();
    private static JSEvaluator jsEvaluator;

    private WebView webView;
    private Evaluator evaluator;

    // <editor-fold desc="Private API">

    public static Observable<JSEvaluator> getInstance(Context context) {
        return getInstance("", context);
    }

    public static Observable<JSEvaluator> getInstance(String url, Context context) {

        if (!initializedJsEvaluator.getAndSet(true)) {
            // We need to start initializing the singleton
            new JSEvaluator(url, new WebView(context), evaluator -> jsEvaluatorSubject.onNext(evaluator));
        } else if (jsEvaluator != null) {
            // We've already initialized the singleton
            return Observable.just(jsEvaluator);
        }
        // Another caller started initialization, await its result
        return jsEvaluatorSubject.first();
    }

    private JSEvaluator(String url, WebView webView, JSEvaluatorCallback callback) {
        this.webView = webView;
        this.evaluator = new LegacyEvaluator(webView);
        // Not sure I want to have two different javascript pipelines
        // especially bc there seems to be some inconsistent behavior:
        // https://github.com/evgenyneu/js-evaluator-for-android/issues/4
//        this.evaluator = Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT ?
//                new KitKatEvaluator() : new LegacyEvaluator(webView);
        setupWebView(url, callback);
    }

    private void setupWebView(String url, JSEvaluatorCallback callback) {
        webView.setWillNotDraw(true);
        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);

        webView.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageFinished(WebView view, String url) {
                Timber.d("onPageFinished");
                AndroidSchedulers.mainThread().scheduleDirect(() -> {
                    Timber.d("Notifying callback");
                    jsEvaluator = JSEvaluator.this;
                    callback.onReady(JSEvaluator.this);
                });
            }

            @Override
            public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
                Timber.e("Webview error " + errorCode + " desc " + description);
            }
        });

        Timber.d("Loading html");
        webView.loadUrl(url);
    }

    // </editor-fold desc="Private API">

    // <editor-fold desc="Public API">

    public void reverseGeocode(final double latitude, final double longitude, Evaluator.EvaluatorCallback callback) {
        evaluator.evaluate(webView,
                String.format("return window.coder.reverse(%f, %f);", latitude, longitude),
                callback);
    }

    public void forwardGeocode(final String playaAddress, Evaluator.EvaluatorCallback callback) {
        evaluator.evaluate(webView,
                String.format("return window.coder.forwardAsString(\"%s\");", playaAddress),
                callback);
    }

    // </editor-fold desc="Public API">

    public interface JSEvaluatorCallback {
        void onReady(JSEvaluator evaluator);
    }

    public interface Evaluator {

        interface EvaluatorCallback {
            void onResult(String result);
        }

        void evaluate(WebView webView, String script, EvaluatorCallback callback);
    }
}

File: iBurn/src/main/java/com/gaiagps/iburn/js/JsEvaluator.kt
package com.gaiagps.iburn.js

import android.content.Context
import android.widget.Toast
import com.gaiagps.iburn.BuildConfig
import com.squareup.duktape.Duktape
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import java.util.concurrent.Executors

/**
 * Created by dbro on 6/12/17.
 */

/**
 * Use a single threaded executor to prevent spinning up multiple instances of the JS engine,
 * which is very expensive
 */
private val jsScheduler =
        Schedulers.from(
                Executors.newSingleThreadExecutor()
        )

object JsEvaluator {

    val jsPath = "js/bundle.js"
    private var duktape: Duktape? = null
    private var jsContent: String? = null
    private var geocoder: GeoCoder? = null

    internal interface GeoCoder {
        fun reverse(lat: Double, lon: Double): String
        fun forward(playaAddress: String): String
    }

    fun reverseGeocode(context: Context, lat: Float, lon: Float): Single<String> {

        return Single.just(true)
                .observeOn(jsScheduler)
                .map { ignored ->

                    if (jsContent == null) {
                        Timber.d("Loading JS...")
                        val inStream = context.assets.open(jsPath)
                        jsContent = inStream.bufferedReader().use { it.readText() }
                        inStream.close()
                        Timber.d("Loaded JS")
                    }

                    if (duktape == null) {
                        val startTime = System.currentTimeMillis()
                        Timber.d("Creating Duktape...")
                        val duktape = Duktape.create()
                        duktape.evaluate("var window = this;" + jsContent + "var coder = window.prepare();")
                        this.duktape = duktape
                        val logStr = "Created Duktape in ${(System.currentTimeMillis() - startTime) / 1000f} s"
                        Timber.d(logStr)
                        if (BuildConfig.DEBUG) {
                            AndroidSchedulers.mainThread().scheduleDirect {
                                Toast.makeText(context, logStr, Toast.LENGTH_LONG).show()
                            }
                        }
                    }

                    if (geocoder == null) {
                        Timber.d("Creating Geocoder...")
                        this.geocoder = duktape?.get("coder", GeoCoder::class.java)
                        Timber.d("Created Geocoder")
                    }

                    Timber.d("Geocoding...")
                    // Call into the JavaScript object to decode a string.
                    val playaAddress = geocoder?.reverse(lat.toDouble(), lon.toDouble())
                    Timber.d("Geocode result %s", playaAddress)
                    playaAddress
                }

    }

    fun close() {
        Timber.d("Closing")
        duktape?.close()
        duktape = null
    }
}